src/lib.rs

```diff
@@ -0,0 +1,555 @@
+///     3. node_file: optional csv node file path,
+///     4. edge_file: csv edge file path,
+///     5. separate:  csv files separate where each term is separated with sep,
+///     6. vertices: `vec` for storing nodes which reading from the node file with separate,
+///     7. edges: `vec` for storing edges which reading from the edge file with separate,
+///     8. neighbour_index: `HashMap` for indexing node's neighbours pos in `vec`.
+///
+/// Here are the descriptions for the functions:
+///     1. default(): create a default graph data;
+///     2. with_dir(GraphType::Directed(or Undirected)): the graph constructing is directed or not;
+///     3. with_schema(): identifing the schema of the graph;
+///     4. from_file(edge_file:String,node_file:Option<String>,sep:String): node_file path,
+///             and optional edge file path, both are csv files where each term is separated with sep.
+///     5. build(): return a graph instance according to the iniformations gave above.
+///
+/// # Example
+///
+/// ```
+///  use graph_rust::{Graph, GraphType};
+///  let graph = Graph::default()
+///        .with_dir(GraphType::Directed)
+///        .with_schema(String::from("..."))
+///        .from_file(String::from("data_edges.csv"), None, ",")
+///        .build();
+/// ```
+#[derive(Clone)]
+pub struct Graph<T> {
+    graph_type: GraphType,
+    schema: String,
+    node_file: Option<String>,
+    edge_file: String,
+    separate: String,
+    vertices: Vec<Vertex<T>>,
+    edges: Vec<Edge<T>>,
+    neighbour_index: HashMap<T, (usize, usize)>,
+}
+
+impl<T> Graph<T>
+    where
+        T: Eq + Hash + FromStr + Display + Clone + Copy,
+        <T as FromStr>::Err: Debug,
+{
+    pub fn default() -> Graph<T> {
+        return Graph {
+            graph_type: GraphType::Directed,
+            schema: "".to_string(),
+            node_file: None,
+            edge_file: "".to_string(),
+            separate: "".to_string(),
+            vertices: vec![],
+            edges: vec![],
+            neighbour_index: HashMap::new(),
+        };
+    }
+
+    pub fn with_dir(&mut self, dir: GraphType) -> &mut Graph<T> {
+        self.graph_type = dir;
+        self
+    }
+
+    pub fn with_schema(&mut self, schema: String) -> &mut Graph<T> {
+        self.schema = schema;
+        self
+    }
+
+    pub fn from_file(
+        &mut self,
+        edge_file: String,
+        node_file: Option<String>,
+        sep: &str,
+    ) -> &mut Graph<T> {
+        self.node_file = node_file;
+        self.edge_file = edge_file;
+        self.separate = sep.to_string();
+        self
+    }
+
+    pub fn build(&mut self) -> Graph<T> {
+        //Init node and edge container
+        let mut vertices = HashMap::new();
+        let mut edges = HashMap::new();
+
+        let directed = match self.graph_type {
+            GraphType::Directed => true,
+            GraphType::Undirected => false,
+        };
+
+        let no_node_file = match self.node_file {
```

Comment:
It does not make sense to maintain `node_file` and `edge_file` in Graph structure. After reading the data, they are useless. You can probably construct another class named `GraphReader` to do this instead.

Normally, the data may be very large. Therefore, can you come up with solutions to reader graph file using multi-threads. That is, you can predefine t threads in the `GraphReader` structure, and then use t-1 threads to read files, and the remaining one to construct graph? If using arbitrary t is too hard, at least using 2.